### PR TITLE
Remove dependency of hono-client-adapter on hono-client

### PR DESCRIPTION
**Pull request by sophokles73**
Factored out common exception classes into new module hono-client-core.
This is for #2156 

@b-abel can you also take a look and see if the changes made to the client interfaces work with Kafka?

### CODE REVIEW

**Review by calohmn — APPROVED** [review]
Looks good from my side
(assuming it's no issue OSGI-wise that the created hono-client.jar now contains the added "Embed-Dependency" entry in `META-INF/MANIFEST.MF` and an added `META-INF/maven/org.eclipse.hono/hono-client-core` directory with pom.xml and pom.properties).

**Comment by sophokles73** [comment]
@calohmn no, it shouldn't

**Comment by b-abel** [comment]
@sophokles73 The lifecycle methods imply for me that you can stop and restart a client instance. Is this intended?

**Comment by sophokles73** [comment]
> The lifecycle methods imply for me that you can stop and restart a client instance. Is this intended?

The method are intended to start/stop the underlying connection to the broker. However, I wonder if it would be better to only let implementations (optionally) implement Lifecycle instead of making it mandatory for TelemetrySender and EventSender ...

**Comment by sophokles73** [comment]
@b-abel The `ServiceClient` interface currently extends `Closable` which serves a similar purpose as `Lifecycle.stop()`, so I wonder if we should drop the `Closeable` reference altogether and instead let client implementations (optionally) implement `Lifecycle`. WDYT?
We can then also think about making implementing `ServiceClient` optional as well. Or does it always make sense for `TelemetrySender` and `EventSender` implementations to register health checks?

File: clients/adapter/src/main/java/org/eclipse/hono/adapter/client/telemetry/EventSender.java [hunk 1]
```diff
@@ -44,13 +47,13 @@
      *         The future will be failed with a {@link org.eclipse.hono.client.ServerErrorException} if the data
      *         could not be sent. The error code contained in the exception indicates the
      *         cause of the failure.
-     * @throws NullPointerException tenant ID, device ID or contentType are {@code null}.
+     * @throws NullPointerException if tenant ID, device ID or contentType are {@code null}.
```
[the remark this answers is not in the document]
Comment (sophokles73) [reply clients/adapter/src/main/java/org/eclipse/hono/adapter/client/telemetry/EventSender.java:50]: It is actually not about the authenticated device but the device that the data originates from. The right thing to do would probably be to make sure that `TenantObject` instances always have a tenant ID. For the time being, I will document that the method throws an `IllegalArgumentException` if no tenant ID is set ...

**Comment by b-abel** [comment]
> @b-abel The `ServiceClient` interface currently extends `Closable` which serves a similar purpose as `Lifecycle.stop()`, so I wonder if we should drop the `Closeable` reference altogether and instead let client implementations (optionally) implement `Lifecycle`. WDYT?

Do you want to cast the clients to invoke the lifecycle methods? 
I think a `close` method is quite universal, it is to be expected that clients allocate resources and need to release them again  (client implementations are free to decide what this means in their context). My concern was more that the lifecycle implies that a closed instance can be _re-opened_ later on (at least the documentation says nothing to the contrary). This puts higher requirements on the client implementation which might not be necessary. So, IMHO the interfaces should either extend `Closable` _or_ `Lifecycle`.

> We can then also think about making implementing `ServiceClient` optional as well. Or does it always make sense for `TelemetrySender` and `EventSender` implementations to register health checks?

This is a tough question. Yesterday I designed a health check mechanism for the Kafka client, which might add value. But maybe we should write down clear expectations, in which kind of events the health checks are expected to fail and what will be the consequences.
I assume that you intend to aggregate them in the readiness and liveness checks of the protocol adapters so that a failing liveness check would trigger a pod restart...

**Comment by sophokles73** [comment]
> My concern was more that the lifecycle implies that a closed instance can be re-opened later on (at least the documentation says nothing to the contrary). This puts higher requirements on the client implementation which might not be necessary. So, IMHO the interfaces should either extend Closable or Lifecycle.

We can refine the JavaDoc of the `Lifecyle.stop()` method to make it clear that there is no expectation that the component can be restarted after its `stop()` method has been invoked. But I would rather not support components implementing `Lifecycle` *or* `Closable` but instead only support `Lifecycle`.

**Comment by sophokles73** [comment]
> I assume that you intend to aggregate them in the readiness and liveness checks of the protocol adapters so that a failing liveness check would trigger a pod restart...

Yes, that is the idea. However, it is not so much about the liveness check but more about the readiness check ...

**Comment by b-abel** [comment]
> But I would rather not support components implementing `Lifecycle` _or_ `Closable` but instead only support `Lifecycle`.

That's what I meant :-)

**Comment by b-abel** [comment]
> However, it is not so much about the liveness check but more about the readiness check ...

The readiness check serves a similar purpose as `Lifecyle.start()`. The protocol adapter can use it for its readyness check.
So, if we keep `Lifecycle`, we IMHO could remove the health check methods and the `Closable` interface. WDYT?

**Comment by sophokles73** [comment]
> The readiness check serves a similar purpose as Lifecyle.start().

No, the readiness check is being performed periodically (triggered by an external monitor) whereas the *start* method is being invoked once during startup of the component. In our case the readiness check verifies, if  for example the connection to the service is still open and working. The start method is invoked once only to initially establish the connection ...

**Comment by sophokles73** [comment]
I have pushed some changes which work for the AMQP based implementation. @b-abel can you check, if this works for the Kafka based implementation as well?

**Comment by b-abel** [comment]
> I have pushed some changes which work for the AMQP based implementation. @b-abel can you check, if this works for the Kafka based implementation as well?

You removed `Closeable` as well as `Lifecycle`. The Kafka client needs to be closed at some point. I think this would apply to nearly every client. When implementing a client I would like to see this expectation explicitly in the interface and be sure which method to implement to release ressources. When implementing a protocol adapter I would like not to do type casts to find out how the client is to be closed.

**Comment by sophokles73** [comment]
simply implement `org.eclipse.hono.util.Lifecycle` ...

**Comment by b-abel** [comment]
> So now you are for making implementing `org.eclipse.hono.util.Lifecycle` mandatory, right?

Yes, I was never against it. I am sorry if my comments were not clear enough. I thought we agreed that having `Lifecycle` and `Closeable` mandatory was too much. And I understood that you favor `Lifecycle` over `Closeable`. 

So, now hopefully more clear:
_I am for a mandatory close method_. I care less about its name (`Lifecyle.stop()` is fine as well) or its signature.

**Comment by sophokles73** [comment]
@b-abel Ok, I have changed the service clients to extend Lifecycle now. Hope this works for everybody now ...

**Pull request merged**